### PR TITLE
DOM-35464 convert passed in hardware tier name to hardware tier id

### PR DESCRIPTION
**Pull request by ddl-awroblicky**
still working on adding tests

### CODE REVIEW

File: domino/domino.py
Comment (dknupp) [inline domino/domino.py:822]: AFAIK, hardware tier names are not required to be unique, so this will be an issue. I think this why overriding by name was probably a mistake in the first place.

It's unfortunate, but I think to address this issue, we really need to change the interface of `job_start()` to take hardware ID from the start. At least we know that no customer is using this as is (because it's totally broken, and no one has reported a bug before now).

**Review by dknupp — COMMENTED** [review]
Thanks for the quick response on this. I think we need to get away from the hardware tier name entirely, however, for reasons noted.

Also, note that there have been recent efforts to add unit tests for any new changes to the library. My gut is that some testing could be added here:
https://github.com/dominodatalab/python-domino/blob/master/tests/test_jobs.py

File: domino/domino.py
Comment (dknupp) [reply domino/domino.py]: An alternative to changing the interface would be to accept the name as exists right now, get the ID like you proposed, but throw an error in the event that more than one HWT with the same name is found.

I'm honestly not sure which is the better option.

Comment (ddl-awroblicky) [reply domino/domino.py]: perhaps similar to `on_demand_spark_cluster_properties`, I can document it as deprecated and try the best effort approach and fail if multiple HWT with the same name are found. This way if any customer is using it but not noticing it doesn't work, their code will at least still run.

I can also add a new optional property for HWT id that is recommended to be used instead.

Comment (ddl-awroblicky) [reply domino/domino.py:829]: @dknupp Does this approach seem ok?

Comment (dknupp) [reply domino/domino.py:829]: @ddl-awroblicky -- ah, I saw this was still marked as a draft, so I didn't know if there were more commits coming. I'll try to have a look later this afternoon or tomorrow.

**Comment by dknupp** [comment]
@ddl-awroblicky -- thanks for adding the unit tests!

I'll try to review this today.

File: domino/domino.py
Comment (dknupp) [inline domino/domino.py:435]: You might be able to shorten this line with...
```suggestion
        resolved_hardware_tier_id = hardware_tier_id or self.get_hardware_tier_id_from_name(hardware_tier_name)
```

File: domino/routes.py
Comment (dknupp) [inline domino/routes.py:154]: Can admin endpoints only be called by admin users?

File: tests/test_hardware_tiers.py
Comment (dknupp) [inline tests/test_hardware_tiers.py:16]: This test is creating a project, not a hardware tier. Maybe a copy/paste error?

File: tests/test_hardware_tiers.py
Comment (dknupp) [inline tests/test_hardware_tiers.py:27]: Similarly, this test still seems to be archiving a project, not a hardware tier.

File: tests/test_jobs.py
Comment (dknupp) [inline tests/test_jobs.py:108]: I'm surprised len() isn't throwing an exception here, something like...
```python
>>> num_gen_expr = (n for n in [1,2,3])
>>> len(num_gen_expr)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: object of type 'generator' has no len()
```

File: tests/test_jobs.py
Comment (dknupp) [inline tests/test_jobs.py:106]: This test ran and passed? The way I read the current line, `non_default_hardware_tiers` is being assigned a python generator that will produce a series of Boolean values. It looks like what you want a list comprehension instead, something like...
```suggestion
    non_default_hardware_tiers = [hwt for hwt in hardware_tiers if not hwt["isDefault"]]
```

File: domino/routes.py
Comment (ddl-awroblicky) [reply domino/routes.py:154]: yes that should be the case

if not it is a bug

File: tests/test_jobs.py
Comment (dknupp) [inline tests/test_jobs.py:92]: PEP-8 requires 2 lines between module-level functions.
```suggestion

@pytest.mark.skipif(not domino_is_reachable(), reason="No access to a live Domino deployment")
```

File: tests/test_jobs.py
Comment (dknupp) [inline tests/test_jobs.py:121]: PEP-8 requires 2 lines between module-level functions.
```suggestion

@pytest.mark.skipif(not domino_is_reachable(), reason="No access to a live Domino deployment")
```

File: tests/test_jobs.py
Comment (dknupp) [inline tests/test_jobs.py:106]: PEP-8 requires 2 lines between module-level functions.
```suggestion

# deprecated but ensuring it still works for now
```

File: tests/test_jobs.py
Comment (dknupp) [inline tests/test_jobs.py:119]: The way this is written, it leaves open the possibility of a falsely passing test. If `len(non_default_hardware_tiers) == 0`, then the actual test never runs -- but it will still appear to be "passing." Reverse the logic and `xfail()` if the test can't run.

```suggestion
    if len(list(non_default_hardware_tiers)) > 0:
        pytest.xfail("No non-default hardware tiers found: cannot run test")

    override_hardware_tier_name = non_default_hardware_tiers[0]["hardwareTier"]["name"]
    job_status = default_domino_client.job_start_blocking(command="main.py",
                                                          hardware_tier_name=override_hardware_tier_name)
    assert job_status['statuses']['isCompleted'] is True
    job_red = default_domino_client.job_runtime_execution_details(job_status['id'])
    assert job_red["hardwareTier"] == override_hardware_tier_name
```

File: tests/test_jobs.py
Comment (dknupp) [inline tests/test_jobs.py:99]: This should already be a list, so the call to `list()` here is redundant.

File: tests/test_jobs.py
Comment (dknupp) [inline tests/test_jobs.py:114]: This should already be a list, so no need to call `list()`.

**Review by dknupp — COMMENTED** [review]
There is a test regression here. On master, if I run the tests which do not require a live host (i.e., pure unit test), they pass.
```
(venv) /path/to/dominodatalab/python-domino(master|…) % pytest -sv tests/test_jobs.py

[...]

tests/test_jobs.py::test_job_status_completes_with_default_params PASSED
tests/test_jobs.py::test_job_status_ignores_RequestException_and_times_out PASSED
tests/test_jobs.py::test_job_status_without_ignoring_exceptions PASSED
```
However, if I checkout your branch and run them, they fail.
```
(venv) /path/to/dominodatalab/python-domino(fix_hardware_tier_name_override↑1|…) % pytest -sv tests/test_jobs.py

[...]

tests/test_jobs.py::test_job_status_completes_with_default_params FAILED
tests/test_jobs.py::test_job_status_ignores_RequestException_and_times_out FAILED
tests/test_jobs.py::test_job_status_without_ignoring_exceptions FAILED
```
I'm trying to dig into why this is happening.

In the meantime, here are some other minor changes that need to be made.

**Review by dknupp — APPROVED** [review]
Thanks @ddl-awroblicky!

**Comment by dknupp** [comment]
@ddl-alexpanin -- still needs code owner approval.

File: tests/test_jobs.py
Comment (ddl-alexpanin) [reply tests/test_jobs.py:144]: @ddl-awroblicky can you update the `CHANGELOG.md`?

**Review by ddl-alexpanin — APPROVED** [review]
LGTM, please update CHANGELOG.md and can merge then.

**Pull request merged**
